README.md
# board_to_fen


Python package that converts digital chessboard image into Forsyth-Edwards notation (FEN) notation

[![Downloads](https://static.pepy.tech/personalized-badge/board-to-fen?period=total&units=none&left_color=purple&right_color=blue&left_text=downloads)](https://pepy.tech/project/board-to-fen)
[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)
[![PyPI](https://img.shields.io/pypi/v/board_to_fen)](https://pypi.org/project/board_to_fen/)
[![GitHub last commit](https://img.shields.io/github/last-commit/mcdominik/board_to_fen)](https://github.com/mcdominik/board_to_fen)

### Installation
board_to_fen is available on PyPI:
```
$ pip3 install board_to_fen
```

### Quick Start
```python
from board_to_fen.predict import get_fen_from_image_path

print(get_fen_from_image_path(PATH_TO_CHESSBOARD_IMAGE))
```

or, if you want you can load image object by yourself:

```python
from PIL import Image
from board_to_fen.predict import get_fen_from_image

img = Image.open(PATH_TO_CHESSBOARD_IMAGE)

print(get_fen_from_image(img))
```


**Note:** *The package uses tensorflow+keras API. 
They are pretty heavy.*

### Customization

get_fen_from_image_path takes has 3 arguments:

- image_path [required]
- end_of_row '/' by default 
- black_view False by default -> set True if chessboard is provided from black player perspective


### Web version (currently may not work)
Available at: https://board2fen.bieda.it


### Training
For training You would probably want to download the source code by cloning the repository:
```
$ git clone https://github.com/mcdominik/board_to_fen.git
```
Download training data from:<br>
I will supply url for data in the future

In the main repository dir, run
```
$ python3 ./board_to_fen/train_model.py
```

### Version history

- january 2023
     - versions 0.0.17-25
     - added simple board validation
     - bug fixes

## Warnings
- Image has to be provided in neutral angle (white or black player's perspective).
- Image has to be square (~3% tolerance depending on image resolution).
- Image can't contain paddings, board borders etc. other than 64 squares (with pieces) itself.


### References:
https://www.kaggle.com/datasets/koryakinp/chess-positions


CHANGELOG.txt
===============================
november 2022
    version 0.0.0-0.0.13
     - fix problems related with publishing on pypi
================================
december 2022
    version 0.0.14
    - add automatic resizing (now image can have any size)
===============================
january 2023
    version 0.0.14-0.0.16
     - fix problems related with github actions
================================
january 2023
    version 0.0.17-25
     - add simple board validation
     - bug fixes
================================
================================
february 2023
    version 0.1.0-0.1.1
     - migratation from cv2 to PIL
     - new function for direct image object load
     - add simple tests
     - bug fixes
================================

test_predict.py
import pytest
from board_to_fen.predict import get_fen_from_image, get_fen_from_image_path
from PIL import Image


img_path = "./test_images/test_image1.png"

class TestNetwork:
    def test_path_prediction(self):
        assert get_fen_from_image_path(img_path) == "rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR"
    def test_object_prediction(self):
        img = Image.open(img_path)
        assert get_fen_from_image(img) == "rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR"


board_to_fen/predict.py
import io
from PIL import Image
from .KerasNeuralNetwork import KerasNeuralNetwork
from .utils import Decoder_FEN, Tiler
try:
    import importlib.resources as pkg_resources
except ImportError:
    # Try backported to PY<37 `importlib_resources`.
    import importlib_resources as pkg_resources
from board_to_fen import saved_models


def get_fen_from_image_path(image_path, end_of_row='/', black_view=False) -> str:
    image = Image.open(image_path)
    decoder = Decoder_FEN()
    net = KerasNeuralNetwork()
    f = pkg_resources.open_text(saved_models, 'november_model')
    net.load_model(f.name)
    tiler = Tiler()
    tiles = tiler.get_tiles(img=image)
    predictions = net.predict(tiles=tiles)
    fen = decoder.fen_decode(squares=predictions, end_of_row=end_of_row, black_view=black_view)
    return fen

def get_fen_from_image(image, end_of_row='/', black_view=False) -> str:
    decoder = Decoder_FEN()
    net = KerasNeuralNetwork()
    f = pkg_resources.open_text(saved_models, 'november_model')
    net.load_model(f.name)
    tiler = Tiler()
    tiles = tiler.get_tiles(img=image)
    predictions = net.predict(tiles=tiles)
    fen = decoder.fen_decode(squares=predictions, end_of_row=end_of_row, black_view=black_view)
    return fen

board_to_fen/KerasNeuralNetwork.py
import os
from contextlib import redirect_stderr
with redirect_stderr(open(os.devnull, "w")):
    from keras import models, layers
import numpy as np


class KerasNeuralNetwork:
    def __init__(self) -> None:
        self.CATEGORIES = ["bishop_black", "bishop_white","empty","king_black","king_white","knight_black", "knight_white","pawn_black", "pawn_white", "queen_black","queen_white", "rook_black","rook_white"]
        self.predictions = []
        self.model = models.Sequential()
        self.model.add(layers.Conv2D(50, (3,3), activation='relu', input_shape=(50,50,3)))
        self.model.add(layers.MaxPooling2D((2,2)))
        self.model.add(layers.Conv2D(100,(3,3), activation='relu'))
        self.model.add(layers.MaxPooling2D((2,2)))
        self.model.add(layers.Conv2D(100, (3,3), activation='relu'))
        self.model.add(layers.Flatten())
        self.model.add(layers.Dense(100, activation='relu'))
        self.model.add(layers.Dense(13, activation='softmax'))
        self.model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])

    def fit(self, train_images, train_labels, test_images, test_labels, num_of_epochs=2, batch_size=32):
        self.model.fit(train_images, train_labels, epochs=num_of_epochs, verbose=1, validation_data=(test_images, test_labels), batch_size=batch_size)

    def save(self, path='.') -> None:
        self.model.save(path)

    def evaluate(self, test_images, test_labels):
        loss, accuracy = self.model.evaluate(test_images, test_labels)
        print(f"accuracy:{accuracy}")
        print(f"loss:{loss}")

    def load_model(self, path):
        self.model = models.load_model(path)
    
    def predict(self, tiles) -> list:
        for image in tiles:
            image = np.array(image)
            image = np.reshape(image, ((1, 50, 50, 3)))
            prediction = self.model.predict(image, verbose=0)
            index = np.argmax(prediction)
            self.predictions.append(self.CATEGORIES[index])
        return self.predictions

board_to_fen/utils.py
import os
import numpy as np
import random
import re
from PIL import Image
from itertools import product


LEGEND = {
    'pawn_white' : 'P',
    'pawn_black' : 'p',
    'knight_white' : 'N',
    'knight_black' : 'n',
    'bishop_white' : 'B',
    'bishop_black' : 'b',
    'rook_white' : 'R',
    'rook_black' : 'r',
    'queen_white' : 'Q',
    'queen_black' : 'q',
    'king_white' : 'K',
    'king_black' : 'k',
    'empty':'1'
}

class Decoder_FEN():
    def __init__(self) -> None:
        pass

    def _squeeze(self, input) -> str:
        filtered = re.sub('11111111', '8', input)
        filtered = re.sub('1111111', '7', filtered)
        filtered = re.sub('111111', '6', filtered)
        filtered = re.sub('11111', '5', filtered)
        filtered = re.sub('1111', '4', filtered)
        filtered = re.sub('111', '3', filtered)
        squeezed = re.sub('11', '2', filtered)
        return squeezed

    def _simple_validator(self, squares):
        ''' ensures that model:
            - found only one white king
            - found only one black king
            - number of pieces is not greater than 32

        Args:
        figures: list of figures before decoding
        '''
        count_king_white = 0
        count_king_black = 0
        figures = 0

        for square in squares:
            if square == 'king_white':
                count_king_white+=1
            if square == 'king_black':
                count_king_black+=1
            if square != 'empty':
                figures +=1
        if ((count_king_black != 1) or (count_king_white != 1) or ((figures) > 32)):
            return 'invalid'

    def fen_decode(self, squares, end_of_row='/', black_view=False) -> str:
        if (self._simple_validator(squares) == 'invalid'):
            return 'Model can\'t find valid chessboard layout'
        long_fen = ''
        for i, square in enumerate(squares):
            if i % 8 == 0 and i>0:
                long_fen+=end_of_row
            long_fen+=LEGEND[square]
            fen = self._squeeze(long_fen)
            if black_view:
                fen = fen[::-1]
        return fen


class DataFetcher:
    def __init__(self) -> None: 
        self.CATEGORIES = ["bishop_black", "bishop_white","empty","king_black","king_white","knight_black", "knight_white","pawn_black", "pawn_white", "queen_black","queen_white", "rook_black","rook_white"]        
        self.data = []
        self.images = []
        self.labels = []

    def fetch_and_shuffle(self, data_dir):
        for category in self.CATEGORIES:
            path = os.path.join(data_dir, category)
            class_num = self.CATEGORIES.index(category)
            for filename in os.listdir(path):
                try:
                    img = Image.open(os.path.join(path,filename)).convert('RGB')
                    ima = np.array(img)
                    #consider division /255
                    self.data.append([img/255, class_num])
                except Exception as e:
                    pass
        random.shuffle(self.data)
        self._create_labels_and_images()

    def _create_labels_and_images(self):
        for features, label in self.data:
            self.images.append(features)
            self.labels.append(label)
        self.images = np.array(self.images).reshape(-1, 50, 50, 3)
        self.labels = np.array(self.labels)

    def get_train_test(self, split=0.85):
        pivot = int(split * len(self.images))
        train_images = self.images[:pivot]
        train_labels = self.labels[:pivot]
        test_images = self.images[pivot:]
        test_labels = self.labels[pivot:]
        return train_images, train_labels, test_images, test_labels


class Tiler:
    def __init__(self) -> None:
        self.tile_images = []
        pass

    def get_tiles(self, img, d=50) -> list:
        img = img.resize((400,400), Image.NEAREST).convert('RGB')
        w, h = img.size
        grid = product(range(0, h-h%d, d), range(0, w-w%d, d))
        for i, j in grid:
            box = (j, i, j+d, i+d)
            self.tile_images.append(img.crop(box))
        return self.tile_images
